lib/async_hooks.js

```diff
@@ -0,0 +1,486 @@
+  tmp_async_hook_fields[kDestroy] = async_hook_fields[kDestroy];
+}
+
+
+// Then restore the correct hooks array in case any hooks were added/removed
+// during hook callback execution.
+function restoreTmpHooks() {
+  active_hooks_array = tmp_active_hooks_array;
+  async_hook_fields[kInit] = tmp_async_hook_fields[kInit];
+  async_hook_fields[kBefore] = tmp_async_hook_fields[kBefore];
+  async_hook_fields[kAfter] = tmp_async_hook_fields[kAfter];
+  async_hook_fields[kDestroy] = tmp_async_hook_fields[kDestroy];
+
+  tmp_active_hooks_array = null;
+  tmp_async_hook_fields = null;
+}
+
+
+function createHook(fns) {
+  return new AsyncHook(fns);
+}
+
+
+function currentId() {
+  return async_uid_fields[kCurrentAsyncId];
+}
+
+
+function triggerId() {
+  return async_uid_fields[kCurrentTriggerId];
+}
+
+
+// Embedder API //
+
+class AsyncEvent {
+  constructor(type, triggerId) {
+    this[async_id_symbol] = ++async_uid_fields[kAsyncUidCntr];
+    // Do this before early return to null any potential kInitTriggerId.
+    if (triggerId === undefined)
+      triggerId = initTriggerId();
+    // If a triggerId was passed, any kInitTriggerId still must be null'd.
+    else
+      async_uid_fields[kInitTriggerId] = 0;
+    this[trigger_id_symbol] = triggerId;
+
+    // Return immediately if there's nothing to do.
+    if (async_hook_fields[kInit] === 0)
+      return;
+
+    if (typeof type !== 'string' || type.length <= 0)
+      throw new TypeError('type must be a string with length > 0');
+    if (!Number.isSafeInteger(triggerId) || triggerId < 0)
+      throw new RangeError('triggerId must be an unsigned integer');
+
+    for (var i = 0; i < active_hooks_array.length; i++) {
+      if (typeof active_hooks_array[i][init_symbol] === 'function') {
+        runInitCallback(active_hooks_array[i][init_symbol],
+                        this[async_id_symbol],
+                        type,
+                        triggerId,
+                        this);
+      }
+    }
+  }
+
+  emitBefore() {
+    emitBeforeS(this[async_id_symbol], this[trigger_id_symbol]);
+    return this;
+  }
+
+  emitAfter() {
+    emitAfterS(this[async_id_symbol]);
+    return this;
+  }
+
+  emitDestroy() {
+    emitDestroyS(this[async_id_symbol]);
+    return this;
+  }
+
+  asyncId() {
+    return this[async_id_symbol];
+  }
+
+  triggerId() {
+    return this[trigger_id_symbol];
+  }
+}
+
+
+function asyncIdScope(asyncId, cb) {
+  // Store the async id now to make sure the stack is still good when the ids
+  // are popped off the stack.
+  const prevId = currentId();
+  pushAsyncIds(asyncId, prevId);
+  try {
+    cb();
+  } finally {
+    popAsyncIds(asyncId);
+  }
+}
+
+
+// Sensitive Embedder API //
+
+// Increment the internal id counter and return the value. Important that the
+// counter increment first. Since it's done the same way in
+// Environment::new_async_uid()
+function newUid() {
+  return ++async_uid_fields[kAsyncUidCntr];
```

Comment:
sort of... `asyncId` refers to an id that's been assigned to a resource instance, and should never change for that resource. `asyncUid` refers to the pool of ids that haven't been assigned yet, and whose fields will change. e.g. `async_uid_fields` is an array of `asyncId`s that all represent something at a given moment, but the fields of `async_uid_fields` will change constantly.

But this evolved on its own and I'm not adverse to changing their name.